### DISCUSSION TITLE
Any way to forward https to localhost, not hosted on github?

### DISCUSSION THREAD

**Discussion by awmcclain** [community/community#10560] [Codespaces]
We have a Svelte webapp that is serving over https using a self-signed cert in a codespace, and we have our hosts file on our host computers set up to use a virtual domain for localhost.

Vscode successfully detects the port and forwards it, but to a unique address on the GitHub domain. Can we change that so it forwards to 127.0.0.1? Since it’s not on our domain we have issues with with callbacks, cookies, and OWASP headers.

**Comment by joshspicer** [comment]
Which client are you using (web or desktop)?  I suggest connecting to your codespace via the **desktop** VS Code client, by default for private ports we bind to localhost instead of generating that GitHub port forwarding address.  

You can then edit your hosts file on your machine to work with your self-signed cert.  We do this for internal codespaces development, so it's definitely possible 👍.  We run a barebones nginx docker container to proxy the port-forwarded connection with our test SSL cert (but you may have a different approach that works equally well).

Some info in the docs in case you didn't see them: https://docs.github.com/en/codespaces/developing-in-codespaces/forwarding-ports-in-your-codespace

Note that you need to set your port to `private`.  If you set it to `public`, we need to forward it via a github.dev URI so others can access it :)

![image](https://user-images.githubusercontent.com/23246594/151882577-69a0ca46-13bd-4bd6-bc3c-e081243bc266.png)

**Reply by awmcclain** [reply]
I'm using desktop.

Switching the port protocol to https -- even if it's private -- stops forwarding to localhost and instead hosts the URL on githubpreview.dev

**Reply by awmcclain** [reply]
In our case we have our self signed cert available _inside_ our web server docker instance. 

When we developed locally, we could just serve via docker and then connect via our adjusted hosts file since our local docker was forwarding directly to the host.

We can workaround by serving over http and spinning up a reverse proxy on the host machine, but then that's an extra command to run on host machine.

**Reply by joshspicer** [reply]
Hmm yea I see.  I understand the desire for it to work exactly how it does locally.  That said, personally I think a quick reverse proxy in a docker container on the host may be the most simple approach if you _need_ to use your own self-signed cert (and can't permit the `githubpreview.dev` cert during development).

---

For some dev work i've used a script similar to this (in the example proxying `github.localhost`) to a codespace listening on localhost.

```bash
#!/bin/bash

CONTAINER_NAME=codespaces-nginx

mkdir ./ssl 2> /dev/null;

echo -e "\n* Installing the root CA..\n"

mkcertBin=./bin/mkcert-linux
if [[ "$OSTYPE" == "darwin"* ]]; then
    mkcertBin=./bin/mkcert-darwin
fi

$mkcertBin -install

echo -e "\n* Creating certs..\n"

$mkcertBin -cert-file ./ssl/cert.pem -key-file ./ssl/key.pem github.localhost '*.github.localhost'

# start container
docker_state=$(docker info >/dev/null 2>&1)
if [[ $? -ne 0 ]]; then
    echo -e "\n** Docker is not running, please start docker deamon first."
    exit 1
fi

# docker is running

echo -e "* Starting nginx as $CONTAINER_NAME..\n"

docker rm -f $CONTAINER_NAME &> /dev/null

docker run --name $CONTAINER_NAME -p 443:443 -v $(pwd):/etc/nginx:ro -d nginx

if [ ! "$(docker ps -q -f name=$CONTAINER_NAME)" ]; then
    echo -e "! Could not start the DNS server docker container, terminating.."
    exit 1
fi

echo -e "\n* Done."
```

With boiler plate nginx config - specifically under `./servers` this nginx config

```nginx
log_format cs_log '$remote_addr - $remote_user [$time_local] $status '
                '"$request" $body_bytes_sent "$http_referer" '
                '"$http_user_agent" "http_x_forwarded_for"';


server {
  access_log /var/log/nginx/access.log cs_log;
  client_max_body_size 50M;
  listen 443 ssl;
  listen [::]:443 ssl;

  root ./;
  server_name codespaces.github.localhost ~^(www\.)?codespace\-[\w-]+\.github\.localhost$;

  ssl_certificate ssl/cert.pem;
  ssl_certificate_key ssl/key.pem;

  location / {
    proxy_intercept_errors off;
    proxy_max_temp_file_size 0;
    proxy_redirect off;
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-Proto $scheme;

    if (-f $request_filename/index.html) {
      rewrite (.*) $1/index.html break;
    }

    if (-f $request_filename.html) {
      rewrite (.*) $1.html break;
    }

    if (!-f $request_filename) {
      proxy_pass "http://host.docker.internal:5000";
      break;
    }
  }
}
```

**Reply by awmcclain** [reply]
Caddy works really well for this too, it's a one liner:

Caddyfile
```
{
	local_certs
}

local.mydomain.com

reverse_proxy 127.0.0.1:8080
```
`caddy run`

It's less about the cert, though, and more about the host, since we'd need to configure auth callbacks for every single codespace (because each domain is different) (not to mention OWASP headers, etc)

**Reply by joshspicer** [reply]
Spoke with a teammate of mine more familiar with this, and it sounds you may be able to accomplish this by toggling the following setting in the GitHub Codespaces vscode extension:

<img width="758" alt="Screen Shot 2022-02-01 at 8 44 39 AM" src="https://user-images.githubusercontent.com/23246594/151983722-db860000-c9e4-456b-a4fa-5c132594cc87.png">

**Reply by awmcclain** [reply]
This is perfect! Thank you! It'd be fantastic to add this instruction to https://docs.github.com/en/codespaces/developing-in-codespaces/forwarding-ports-in-your-codespace#forwarding-a-port

**Reply by joshspicer** [reply]
> It'd be fantastic to add this instruction

Agreed! I'll take care of that :)  Thanks for working through this with me

**Comment by Syammed2429** [comment]
Use forward port and better you use it through vscode, it auto sets
